diffkemp/building/cc_wrapper_template.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
+"""
+Template for C compiler wrapper used in build stage of DiffKemp
+
+It's written to be either run with RPython to generate a native code
+executable, or with Python to modify itself into a Python script.
+Because of that it has to run under both Python 2 and Python 3 and some of
+Python library functions are redefined here to be RPython-compatible.
+"""
+# TODO: Decompose simultaneous compiling and linking to get inidividual .ll
+#       files out of it; support other GCC languages than C (Fortran, C++)
+import os
+import sys
+
+
+wrapper_env_vars = {
+    'db_filename': 'DIFFKEMP_WRAPPER_DB_FILENAME',
+    'clang': 'DIFFKEMP_WRAPPER_CLANG',
+    'clang_append': 'DIFFKEMP_WRAPPER_CLANG_APPEND',
+    'clang_drop': 'DIFFKEMP_WRAPPER_CLANG_DROP',
+    'debug': 'DIFFKEMP_WRAPPER_DEBUG'
+}
+
+
+def execl(file, args):
+    # Note: execl from os is not RPython-compatible
+    if os.path.sep in file:
+        os.execv(file, args)
+        return
+    if 'PATH' in os.environ.keys():
+        envpath = os.environ['PATH']
+    else:
+        envpath = os.defpath
+    path = envpath.split(os.pathsep)
+    saved_exc = None
+    last_exc = None
+    for directory in path:
+        fullname = os.path.join(directory, file)
+        try:
+            os.execv(fullname, args)
+        except OSError as e:
+            last_exc = e
+            if saved_exc is None:
+                saved_exc = e
+    if saved_exc is not None:
+        raise saved_exc
+    raise last_exc
+
+
+class CalledProcessError(OSError):
+    pass
+
+
+def check_call(file, args):
+    pid = os.fork()
+    if pid == 0:
+        execl(file, args)
+    _, status = os.waitpid(pid, 0)
+    if status != 0:
+        raise CalledProcessError()
+
+
+def wrapper(argv):
+    # Get arguments from environmental variables
+    db_filename = os.environ.get(wrapper_env_vars['db_filename'])
+    clang = os.environ.get(wrapper_env_vars['clang'])
+    append = os.environ.get(wrapper_env_vars['clang_append']).split(',')
+    drop = os.environ.get(wrapper_env_vars['clang_drop']).split(',')
+    debug = os.environ.get(wrapper_env_vars['debug'])
+
+    if (db_filename is None or clang is None or append is None or
+            drop is None or argv is None):
+        return 1
+
+    db = []
+    # TODO: Allow custom llvm-link and llvm-dis
+    llvm_link = 'llvm-link'
+    llvm_dis = 'llvm-dis'
+
+    # Run GCC
+    argv[0] = 'gcc'
```

Comment:
Why GCC? What if the original command runs clang? Don't we want just to run the original here?